### PR TITLE
Block producer refactor and chain bus integration

### PR DESCRIPTION
**Pull request by leventeliu**
In this new design, we use an immutable database to keep track of the chain state upon last irreversible block. Each optional brach keeps a runtime cache of its local state. Branches will grow, fork, or switch between one each other, driven by block producing and synchronization, and transactions will apply to the branch states.

Theoretically, block producers should maintain an eventual consistency chain given that the preset confirmation meets BFT requirement.

Chain bus model is added to chain structure for further use.

### CODE REVIEW

File: types/issuekeys.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/CovenantSQL/CovenantSQL/blockproducer/interfaces"
+	"github.com/CovenantSQL/CovenantSQL/crypto"
+	"github.com/CovenantSQL/CovenantSQL/crypto/asymmetric"
+	"github.com/CovenantSQL/CovenantSQL/crypto/verifier"
+	"github.com/CovenantSQL/CovenantSQL/proto"
+)
+
+//go:generate hsp
+
+type MinerKey struct {
```
Comment (covenantdb) [inline types/issuekeys.go:30]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type MinerKey should have comment or be unexported

File: types/issuekeys.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	Miner         proto.AccountAddress
+	EncryptionKey string
+}
+
+type IssueKeysHeader struct {
```
Comment (covenantdb) [inline types/issuekeys.go:36]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type IssueKeysHeader should have comment or be unexported

File: xenomint/types/block.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ func (h *SignedBlockHeader) Verify() error {
 	return h.DefaultHashSignVerifierImpl.Verify(&h.BlockHeader)
 }
 
-// Block defines a block including a signed block header and its query list.
+// BPBlock defines a block including a signed block header and its query list.
```
Comment (covenantdb) [inline xenomint/types/block.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
comment on exported type Block should be of the form "Block ..." (with optional leading article)

File: blockproducer/storage.go [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+func updateProvider(profile *types.ProviderProfile) storageProcedure {
+	var (
+		enc *bytes.Buffer
+		err error
+	)
+	if enc, err = utils.EncodeMsgPack(profile); err != nil {
+		return errPass(err)
+	}
+	return func(tx *sql.Tx) (err error) {
+		_, err = tx.Exec(`INSERT OR REPLACE INTO "provider" VALUES (?, ?)`,
```
Comment (auxten) [inline blockproducer/storage.go]: Put column names in the query is better.

File: blockproducer/chain.go [hunk 5]
```diff
@@ -274,497 +292,696 @@ func (c *Chain) checkBlock(b *pt.Block) (err error) {
...
+	irre = head.head.lastIrreversible(c.comfirms)
+	newIrres = irre.fetchNodeList(c.lastIrre.count)
+
+	// Apply irreversible blocks to create dirty map on immutable cache
+	//
+	// TODO(leventeliu): use old metaState for now, better use separated dirty cache.
+	for _, b := range newIrres {
+		for _, tx := range b.block.Transactions {
+			if err := c.immutable.apply(tx); err != nil {
+				log.WithError(err).Fatal("Failed to apply block to immutable database")
```
Comment (auxten) [inline blockproducer/chain.go]: Error or Fatal strings should not be capitalized

File: blockproducer/chain.go [hunk 6]
```diff
@@ -45,224 +49,238 @@ var (
...
-	if err != nil {
-		return nil, err
+		address:     addr,
+		genesisTime: cfg.Genesis.SignedHeader.Timestamp,
+		period:      cfg.Period,
+		tick:        cfg.Tick,
+
+		peers:      cfg.Peers,
+		nodeID:     cfg.NodeID,
+		comfirms:   uint32(m),
```
Comment (auxten) [inline blockproducer/chain.go]: m is a float64, uint32(m) may cause some vuln problems.

**Review by auxten — CHANGES_REQUESTED** [review]

File: blockproducer/chain.go (hunk 6 above)
Comment (xq262144) [reply blockproducer/chain.go]: Spelling mistake: `comfirms`  -> `confirms`

File: blockproducer/config.go [hunk 7]
```diff
@@ -30,21 +30,22 @@ const (
...
 
 	DataFile string
 
 	Server *rpc.Server
 
-	Peers  *proto.Peers
-	NodeID proto.NodeID
+	Peers            *proto.Peers
+	NodeID           proto.NodeID
+	ComfirmThreshold float64
```
Comment (xq262144) [inline blockproducer/config.go]: Typo here.

File: types/init_service_type.go [hunk 8]
```diff
@@ -31,12 +31,13 @@ type InitService struct {
...
-	Memory        uint64                 // reserved memory in bytes
-	LoadAvgPerCPU uint64                 // max loadAvg15 per CPU
-	EncryptionKey string                 `hspack:"-"` // encryption key for database instance
+	TargetMiners        []proto.AccountAddress // designated miners
+	Node                uint16                 // reserved node count
+	Space               uint64                 // reserved storage space in bytes
+	Memory              uint64                 // reserved memory in bytes
+	LoadAvgPerCPU       uint64                 // max loadAvg15 per CPU
+	EncryptionKey       string                 `hspack:"-"` // encryption key for database instance
+	EventualConsistency bool                   `hspack:"-"` // TODO, eventual consistency test, need later refactor
```
Comment (auxten) [inline types/init_service_type.go]: ConsistencyLevel float64 may be better

Comment (xq262144) [reply types/init_service_type.go]: Ok, good idea.

Comment (leventeliu) [reply types/init_service_type.go]: This is a TODO from @xq262144 , which is currently implemented by calling `chain.Query` directly if the flag is set, i.e., the weakest eventual consistency. Need some more tinking in `kayak` later.

**Review by auxten — APPROVED** [review]

**Review by xq262144 — APPROVED** [review]
Approve

**Pull request merged**